External task comms

bbickerton opened the pull request:
## Issues affected/resolved

(See https://github.com/blog/1506-closing-issues-via-pull-requests)
Resolves #142
## Description:

Adds ability for providers to make requests serviced by other coordinators, with the requests and responses proxied through the providers' coordinator. For example, given a setup where `Coordinator1<->ProviderA` and `Coordinator2<->ProviderB`, this will allow `ProviderA` to use `Provider.TaskZ`, where the request and response will be proxied through `Coordinator1<->Coordinator2`. `ProviderA` never talks to anything other than `Coordinator1` and `ProviderB` never talks to anything other than `Coordinator2`.

External requests are controlled via the `acomm.Request.TaskURL` property. Coordinators will proxy based on whether that is present or not.

Additionally, cleaned up the function signature for creating a new `acomm.Request` as it was getting quite long and disbaled tcp keepalives for coordinators to avoid hanging on shutdown. We'll revisit graceful shutdowns in a separate issue.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cerana/cerana/144)

<!-- Reviewable:end -->

bbickerton committed:
Shorten acomm.NewRequest func signature and add convenient setters

The function signature was getting out of hand. All of the values could
be set directly on the exported properties. Shortened NewRequest to just
take a task (since that's required) and added a few convenience methods
for setting the ResponseHook and StreamURL from a string, and Args from
an interface. All of those could be set directly as well.

Additionally, regenerated all READMEs

bbickerton committed:
Google JSON style

bbickerton committed:
Add TaskURL to the Request

bbickerton committed:
External request proxying based on acomm.Request.TaskURL

bbickerton committed:
Add support for taskURL to coordinator-cli

bbickerton committed:
Proxy StreamURL for external requests

bbickerton committed:
Disable coordinator tcp keepalives stop tracking reqs on initial err

joshie commented:
LGTM

mmlb commented:
Reviewed 34 of 34 files at r1, 3 of 3 files at r2, 3 of 3 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFL0Ud8eAnnUclo7gZ8:-1865455260) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
So a nice thing here could be to do something like http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis. It probably wouldn't make sense if all the `Sets` are going to be used, then again maybe...?

`r := NewRequest("sometask")
r.SetResponseHook
// check err
r.SetStreamURL
// check err

// for each option have to call Set\* and check error

//vs
r := NewRequest("sometask", SetResponseHook("a url"), SetStreamURL("a different url"))

// or maybe make NewRequest and Set*s chainable?
r := NewRequest("sometask").SetResponseHook("a url")....

---

_[acomm/tracker.go, line 407 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL656VPEuiz-p47Jup:-KFL656VPEuiz-p47Juq:-1289135707) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker.go#L407)):_
why don't you pass `r.Body` directly to  `json.Unmarshal`.

---

_[acomm/tracker_test.go, line 66 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL3LfLgxmoNwfL8CnN:-KFL3LfMtachcjJqAvE_:52908632) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker_test.go#L66)):_
I take it `s.Responses` is never closed?

---

_[cmd/coordinator-cli/main.go, line 195 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL1l5Iirkie-7TtVnt:-KFL1l5Iirkie-7TtVnu:81452248) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/cmd/coordinator-cli/main.go#L195)):_
perfect, an example to use in my prev comment:

^ 
vs

```
r := acomm.NewRequest(taskName, acomm.SetResponseHook(responseHook), acomm.SetStreamURL(streamURL), acomm.SetArgs(taskArgs)
if r == nil {
    // welp we got an error
}
```

vs

```
r := acomm.NewRequest(taskName).SetResponseHook(responseHook).SetStreamURL(streamURL).SetArgs(taskArgs)
if r == nil {
    // welp we got an error
}
```

the apparent downside is loss of error value, which could be worked around by checking for `r.Errored()` or something after construction instead of returning `nil`.

---

_[cmd/coordinator-cli/main.go, line 32 [r5]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7tXF3TBePbsP-xNj:-KFL7tXGRd5LSpsGyDdP:1377132517) ([raw file](https://github.com/cerana/cerana/blob/b4e628abc73a0368791405f57d13ac8bcebdc857/cmd/coordinator-cli/main.go#L32)):_
<bikeshed>
`_` -> `-` saves a key stroke per option. Plus `_` in args is eww
</bikeshed>

---

_[coordinator/server.go, line 62 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7RCOb3JfpjHJhVYa:-KFL7RCP6CXDdeSRN1_U:-1883274548) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/coordinator/server.go#L62)):_
unwrap this line, its not common in go and a bit confusing, ditto with similar above.

---

_[coordinator/server.go, line 33 [r7]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL9mE1cifwB8x2RUHH:-KFL9mE1cifwB8x2RUHI:716875727) ([raw file](https://github.com/cerana/cerana/blob/970d523c9f3e31eedff1708250a4a2f1e599703c/coordinator/server.go#L33)):_
This makes more sense in `func init()` since this change affects the whole package/binary. Maybe we should set it only for the http server used below in the mux?

---

_[provider/server.go, line 38 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7a6Fj2yDEBsJBsTO:-KFL7a6Fj2yDEBsJBsTP:1220553242) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/provider/server.go#L38)):_
may be a candidate for api updates...

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFLF1Z2V5Zlgwn5p0Wu:1994994739) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
Hmm. Interesting approach I haven't seen before. Functional options are flexible, but also a bit weird to wrap your head around for people who haven't seen it before. Chaining looks like it'd be tricky to do error handling. @joshie want to weigh in before I make any changes?

---

_[acomm/tracker.go, line 407 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL656VPEuiz-p47Jup:-KFLHRxN55aHl6jg0_V-:-1004557107) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker.go#L407)):_
`json.Unmarshal` wants a `[]byte` while `http.Response.Body` is an `io.ReadCloser`. Is either read and unmarshal or give the body to a `json.NewDecoder` and unmarshal. Two steps either way, and in my opinion the former makes it clearer that this all of the data is a single json object to decode.

---

_[acomm/tracker_test.go, line 66 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL3LfLgxmoNwfL8CnN:-KFLHsgUYhocNoqfL0ww:1798441257) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker_test.go#L66)):_
Correct.

---

_[cmd/coordinator-cli/main.go, line 195 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL1l5Iirkie-7TtVnt:-KFLJkegNyrDtDAmD-K9:-1966699561) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/cmd/coordinator-cli/main.go#L195)):_
Discussion above where the primary comment on this style is.

---

_[cmd/coordinator-cli/main.go, line 32 [r5]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7tXF3TBePbsP-xNj:-KFLI0wL2S-BtwhYWjtE:1642994683) ([raw file](https://github.com/cerana/cerana/blob/b4e628abc73a0368791405f57d13ac8bcebdc857/cmd/coordinator-cli/main.go#L32)):_
I actually want to revisit naming style of command line args, json config file fields, and struct json. (e.g. json should be Google style json camelCase, cli fields should be [some style], and some sort of normalization with viper to make the cli fields and config json fields line up). New Issue: https://github.com/cerana/cerana/issues/149

---

_[coordinator/server.go, line 62 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7RCOb3JfpjHJhVYa:-KFLK0m9I-WOCM42oCXw:117395057) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/coordinator/server.go#L62)):_
Are we just not caring about line length anymore?

---

_[provider/server.go, line 38 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7a6Fj2yDEBsJBsTO:-KFLK73lrZXL8IFo5b73:650295385) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/provider/server.go#L38)):_
Tracker, Request, and Response all could be candidates.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

_[coordinator/server.go, line 33 [r7]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL9mE1cifwB8x2RUHH:-KFLKuKvm4rjjIDQbL6C:-997505445) ([raw file](https://github.com/cerana/cerana/blob/970d523c9f3e31eedff1708250a4a2f1e599703c/coordinator/server.go#L33)):_
Setting it on the server would change the results from "no keepalive on outbound requests" vs "no keepalive on inbound requests". Could be that's what we want. There's an issue to take a deeper look at how graceful stuff is being handled with a note about keepalives: https://github.com/cerana/cerana/issues/148

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFLO3ViZt96MNRx-eFI:1205611790) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
@joshie take a look at another code comment to get some more context/examples. The nice thing about the functional options is that we can get back an error easily, which is a `+` over chaining.

---

_[acomm/tracker.go, line 407 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL656VPEuiz-p47Jup:-KFLN7URngO_yJXmJ6K8:1836279235) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker.go#L407)):_
The problem is that `ioutil.ReadAll` is a bit of an anti-pattern since it will read all of the incoming message and buffer into memory, which can lead to dos, malicious or by mistake. Which should always be streaming.

---

_[coordinator/server.go, line 62 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7RCOb3JfpjHJhVYa:-KFLNqRVViJVqhDZHBnR:82494635) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/coordinator/server.go#L62)):_
never really care too much in go world. The stuff you do in `filepath.Join` helps readability, this here hurts it.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton committed:
Use a json.Decoder instead of ioutil.ReadAll in ProxyExternal

bbickerton committed:
Remove hard wrap on ParseRequestURIs in coordiantor NewServer

bbickerton commented:
Review status: 36 of 38 files reviewed at latest revision, 4 unresolved discussions.

---

_[acomm/tracker.go, line 407 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL656VPEuiz-p47Jup:-KFLOm1zAuhvo3Vqyilj:-132443320) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker.go#L407)):_
Good points. Updated to use a `json.Decoder`. I'm sure there's other places using `ioutil.ReadAll`, but that's for another PR.

---

_[coordinator/server.go, line 62 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL7RCOb3JfpjHJhVYa:-KFLQ2ngRjxoWjUnjQ5E:-1335632205) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/coordinator/server.go#L62)):_
Fair enough. Changed.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[acomm/tracker.go, line 407 [r4]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL656VPEuiz-p47Jup:-KFLQ_gmokSJThXPWlgn:1172325480) ([raw file](https://github.com/cerana/cerana/blob/a33cb1974d8c1c362863fa5f0a258694a1478189/acomm/tracker.go#L407)):_
`s/Which/We/` ... Ack on the other possible `ioutil.ReadAll`s

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

joshie commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFdzh-xrYEY-zYGUZqh:-1594991575) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
Option structs!

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton committed:
Refactor NewRequest to use accept an options struct

bbickerton commented:
Review status: 9 of 38 files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFefiRypnBBEGUPjvhd:286082972) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
Refactored to use an options struct. Kept around the setter functions for convenience.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

joshie commented:
Review status: 9 of 38 files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 33 [r1]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFL0Ud8eAnnUclo7gZ7:-KFegPPotqpLUN4pdXrH:676834233) ([raw file](https://github.com/cerana/cerana/blob/db45789303336c372ec949a18bddb25cba4de2b1/acomm/request.go#L33)):_
LGTM

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 29 of 29 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 52 [r10]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFezGIl48JIwoXskBDy:-KFezGIl48JIwoXskBDz:-1082559756) ([raw file](https://github.com/cerana/cerana/blob/7ace6ac5008caf57831f6480255dc4b7af4959cb/acomm/request.go#L52)):_
why not just make it a struct instead of a pointer then?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 52 [r10]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFezGIl48JIwoXskBDy:-KFf-NlxmdzaphyfF3kM:-303322826) ([raw file](https://github.com/cerana/cerana/blob/7ace6ac5008caf57831f6480255dc4b7af4959cb/acomm/request.go#L52)):_
So we don't need to copy data around

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

mmlb commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 52 [r10]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFezGIl48JIwoXskBDy:-KFfLjhDRb1Y0k8wQNvB:1431632674) ([raw file](https://github.com/cerana/cerana/blob/7ace6ac5008caf57831f6480255dc4b7af4959cb/acomm/request.go#L52)):_
Do you mean at the function call? If so thats a premature optimisation that likely has no impact (not to mention could lead to more garbage for the GC ;) ).

Having a pointer arg over plain struct leads to confusion because now we have to explain that it's basically read-only where a plain struct doesn't need the doc.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 52 [r10]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFezGIl48JIwoXskBDy:-KFincHrF8QEbsAF0Y87:-1154297670) ([raw file](https://github.com/cerana/cerana/blob/7ace6ac5008caf57831f6480255dc4b7af4959cb/acomm/request.go#L52)):_
I guess. I feel like we usually use pointers almost everywhere there's a struct though.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

bbickerton committed:
Use a value intead of pointer for RequestOptions

bbickerton commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[acomm/request.go, line 52 [r10]](https://reviewable.io:443/reviews/cerana/cerana/144#-KFezGIl48JIwoXskBDy:-KFiqcIykIRo-bdnkxoV:257435119) ([raw file](https://github.com/cerana/cerana/blob/7ace6ac5008caf57831f6480255dc4b7af4959cb/acomm/request.go#L52)):_
Changed

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->

mmlb commented:
Reviewed 29 of 29 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/cerana/cerana/144)_

<!-- Sent from Reviewable.io -->
